### PR TITLE
Typescriptify the ClientFactory

### PR DESCRIPTION
**Pull request by Half-Shot**

### CODE REVIEW

File: src/components/client-factory.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+     * will return the *same* client instance.
+     * @param {?string} userId Required. The user_id to scope the client to. A new
+     * client will be created per user ID. If this is null, a client scoped to the
+     * application service *itself* will be created.
+     * @param {Request=} request Optional. The request ID to additionally scope the
+     * client to. If set, this will create a new client per user ID / request combo.
+     * This factory will dispose the created client instance when the request is
+     * resolved.
+     */
+    public getClientAs(userId: string, request: any) {
```
Comment (jaller94) [inline src/components/client-factory.ts:111]:
```suggestion
    public getClientAs(userId?: string, request?: any) {
```

File: src/components/client-factory.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        });
+    };
+
+    /**
+     * Construct a new Matrix JS SDK Client. Calling this twice with the same args
+     * will return the *same* client instance.
+     * @param {?string} userId Required. The user_id to scope the client to. A new
+     * client will be created per user ID. If this is null, a client scoped to the
+     * application service *itself* will be created.
+     * @param {Request=} request Optional. The request ID to additionally scope the
```
Comment (jaller94) [inline src/components/client-factory.ts:106]:
```suggestion
     * @param request The request ID to additionally scope the
```

File: src/components/client-factory.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                // Invoke the callback
+                callback(err, response, body);
+            });
+        });
+    };
+
+    /**
+     * Construct a new Matrix JS SDK Client. Calling this twice with the same args
+     * will return the *same* client instance.
+     * @param {?string} userId Required. The user_id to scope the client to. A new
```
Comment (jaller94) [inline src/components/client-factory.ts:103]:
```suggestion
     * @param userId The user_id to scope the client to. A new
```

Comment (jaller94) [reply src/components/client-factory.ts:103]: I don't understand why it said "Required" though.

File: src/components/client-factory.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+ */
+export class ClientFactory {
+    private clients: { [request_id: string]: { [user_id: string]: any} } = {};
+    private sdk: any;
+    private clientSchedulerBuilder: () => {};
+    private url: string = "";
+    private token: string = "";
+    private botUserId: string= "";
+
+    constructor(opts: { sdk?: any, url: string, token: string, appServiceUserId: string, clientSchedulerBuilder: any}) {
```
Comment (jaller94) [inline src/components/client-factory.ts:44]:
```suggestion
    constructor(opts?: { sdk?: any, url: string, token?: string, appServiceUserId?: string, clientSchedulerBuilder?: any}) {
```

File: src/components/client-factory.ts [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+ * @param {*=} opts.sdk The Matrix JS SDK require() to use.
+ * @param {string=} opts.url The Client-Server base HTTP URL. This must be set
+ * prior to calling getClientAs(). See configure() to set this after instantiation.
+ * @param {string=} opts.token The application service token to use. This must
+ * be set prior to calling getClientAs(). See configure() to set this after
+ * instantiation.
+ * @param {string=} opts.appServiceUserId The application service's user ID. Must
+ * be set prior to calling getClientAs(). See configure() to set this after
+ * instantiation.
+ * @param {function=} opts.clientSchedulerBuilder Optional. A function that
```
Comment (jaller94) [inline src/components/client-factory.ts:32]:
```suggestion
 * @param opts.clientSchedulerBuilder A function that
```

File: src/components/client-factory.ts [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+/**
+ * @constructor
+ * @param {Object} opts Options for this factory
+ * @param {*=} opts.sdk The Matrix JS SDK require() to use.
+ * @param {string=} opts.url The Client-Server base HTTP URL. This must be set
+ * prior to calling getClientAs(). See configure() to set this after instantiation.
+ * @param {string=} opts.token The application service token to use. This must
+ * be set prior to calling getClientAs(). See configure() to set this after
+ * instantiation.
+ * @param {string=} opts.appServiceUserId The application service's user ID. Must
```
Comment (jaller94) [inline src/components/client-factory.ts:29]:
```suggestion
 * @param opts.appServiceUserId The application service's user ID. Must
```

File: src/components/client-factory.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+type LogWrapCallback = (err: Error, response: { statusCode: number }, body: any) => void;
+type OriginalRequest = (opts: any, cb: LogWrapCallback) => void;
+
+/**
+ * @constructor
+ * @param {Object} opts Options for this factory
+ * @param {*=} opts.sdk The Matrix JS SDK require() to use.
+ * @param {string=} opts.url The Client-Server base HTTP URL. This must be set
+ * prior to calling getClientAs(). See configure() to set this after instantiation.
+ * @param {string=} opts.token The application service token to use. This must
```
Comment (jaller94) [inline src/components/client-factory.ts:26]:
```suggestion
 * @param opts.token The application service token to use. This must
```

File: src/components/client-factory.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+*/
+
+type LogWrapCallback = (err: Error, response: { statusCode: number }, body: any) => void;
+type OriginalRequest = (opts: any, cb: LogWrapCallback) => void;
+
+/**
+ * @constructor
+ * @param {Object} opts Options for this factory
+ * @param {*=} opts.sdk The Matrix JS SDK require() to use.
+ * @param {string=} opts.url The Client-Server base HTTP URL. This must be set
```
Comment (jaller94) [inline src/components/client-factory.ts:24]:
```suggestion
 * @param opts.url The Client-Server base HTTP URL. This must be set
```

File: src/components/client-factory.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+limitations under the License.
+*/
+
+type LogWrapCallback = (err: Error, response: { statusCode: number }, body: any) => void;
+type OriginalRequest = (opts: any, cb: LogWrapCallback) => void;
+
+/**
+ * @constructor
+ * @param {Object} opts Options for this factory
+ * @param {*=} opts.sdk The Matrix JS SDK require() to use.
```
Comment (jaller94) [inline src/components/client-factory.ts:23]:
```suggestion
 * @param opts.sdk The Matrix JS SDK require() to use.
```

File: src/components/client-factory.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+type LogWrapCallback = (err: Error, response: { statusCode: number }, body: any) => void;
+type OriginalRequest = (opts: any, cb: LogWrapCallback) => void;
+
+/**
+ * @constructor
+ * @param {Object} opts Options for this factory
```
Comment (jaller94) [inline src/components/client-factory.ts:22]:
```suggestion
 * @param opts Options for this factory
```

File: src/components/client-factory.ts (hunk 4 above)
Comment (jaller94) [reply src/components/client-factory.ts:44]: opts can be optional because of the following line.

File: src/components/client-factory.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        this.clients[reqId][userIdKey] = client;
+
+        return client;
+    };
+
+    /**
+     * Configure the factory for generating clients.
+     * @param {string} baseUrl The base URL to create clients with.
+     * @param {string} appServiceToken The AS token to use as the access_token
+     * @param {string} appServiceUserId The AS's user_id
```
Comment (jaller94) [inline src/components/client-factory.ts:163]:
```suggestion
     * @param baseUrl The base URL to create clients with.
     * @param appServiceToken The AS token to use as the access_token
     * @param appServiceUserId The AS's user_id
```

File: src/components/client-factory.ts [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    constructor(opts: { sdk?: any, url: string, token: string, appServiceUserId: string, clientSchedulerBuilder: any}) {
+        opts = opts || {};
+        this.sdk = opts.sdk || require("matrix-js-sdk");
+        this.clientSchedulerBuilder = opts.clientSchedulerBuilder || function() {};
+        this.configure(opts.url, opts.token, opts.appServiceUserId);
+    }
+
+    /**
+     * Set a function to be called when logging requests and responses.
+     * @param {Function} func The function to invoke. The first arg is the string to
```
Comment (jaller94) [inline src/components/client-factory.ts:53]:
```suggestion
     * @param func The function to invoke. The first arg is the string to
```

File: src/index.ts [hunk 13]
```diff
@@ -13,7 +13,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-module.exports.ClientFactory = require("./components/client-factory");
+export * from "./components/client-factory";
```
Comment (jaller94) [inline src/index.ts:16]: Do I need to do this as well in #185?

Comment (Half-Shot) [reply src/index.ts:16]: Probably yup, do the tests pass without it?

**Pull request merged**